File: Behavioural-Patterns/Observer/Java/Implementations/CustomButton.java
package Implementations;

import java.util.ArrayList;
import java.util.List;

import javax.swing.JButton;

import Interfaces.Observer;
import Interfaces.Subject;

public class CustomButton extends JButton implements Subject {
    private final List<Observer> observers = new ArrayList<>();

    public CustomButton(String text){
        super(text);

        this.addActionListener(e -> notifySubscribers());
    }


    @Override
    public void subscribe(Observer observer) {
        observers.add(observer);
    }

    @Override
    public void unsubscribe(Observer observer) {
        observers.remove(observer);
    }

    @Override
    public void notifySubscribers() {
        System.out.println("Button '" + getText() + "' was clicked. Notifying " + observers.size() + " custom observers.");
        for (Observer observer : observers) {
            observer.update(this);
        }
    }
}

File: Behavioural-Patterns/Observer/C#/Observer/Implementations/CustomButton.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;

public class CustomButton : Button, Subject {
    private readonly List<IObserver> observers = new List<IObserver>();

    public CustomButton()
    {
        this.Click += OnButtonClicked;
    }
    
    private void OnButtonClicked(object sender, EventArgs e)
    {
        NotifySubscribers();
    }

    public void Subscribe(IObserver observer) {
        observers.Add(observer);
    }

    public void Unsubscribe(IObserver observer) {
        observers.Remove(observer);
    }

    public void NotifySubscribers() {
        Console.WriteLine($"Button {this.Text} was clicked. Notifying {observers.Count} custom observers.");
        foreach (IObserver observer in observers) {
            observer.Update(this);
        }
    }
}
